File: app/src/main/java/com/hover/stax/balances/BalancesFragment.java
package com.hover.stax.balances;

import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.cardview.widget.CardView;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

import com.hover.stax.R;
import com.hover.stax.channels.Channel;
import com.hover.stax.databinding.FragmentBalanceBinding;
import com.hover.stax.home.HomeFragment;
import com.hover.stax.home.MainActivity;
import com.hover.stax.navigation.NavigationInterface;
import com.hover.stax.utils.Constants;
import com.hover.stax.utils.UIHelper;
import com.hover.stax.utils.bubbleshowcase.BubbleShowCase;
import com.hover.stax.views.staxcardstack.StaxCardStackView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import timber.log.Timber;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;


public class BalancesFragment extends Fragment implements NavigationInterface {

    final public static String TAG = "BalanceFragment";
    final private String GREEN_BG = "#46E6CC";
    final private String BLUE_BG = "#04CCFC";

    private boolean SHOW_ADD_ANOTHER_ACCOUNT = false;
    final private static int STACK_OVERLAY_GAP = 10;
    final private static int ROTATE_UPSIDE_DOWN = 180;
    private boolean SHOWN_BUBBLE_MAIN_ACCOUNT = false;
    private boolean SHOWN_BUBBLE_OTHER_ACCOUNT = false;

    private BalancesViewModel balancesViewModel;

    private CardView addChannelLink;
    private TextView balanceTitle;
    private boolean balancesVisible = false;

    private RecyclerView balancesRecyclerView;

    private FragmentBalanceBinding binding;
    private StaxCardStackView balanceStack;
    private List<Channel> channelList;
    private BubbleShowCase firstAccountBubble, secondAccountBubble;

    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        balancesViewModel = new ViewModelProvider(requireActivity()).get(BalancesViewModel.class);
        binding = FragmentBalanceBinding.inflate(inflater, container, false);

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        balanceStack = binding.stackBalanceCards;
        setUpBalances();
        setUpLinkNewAccount();
    }

    @Override
    public void onPause() {
        super.onPause();
        if(firstAccountBubble !=null) firstAccountBubble.dismiss();
        if(secondAccountBubble !=null) secondAccountBubble.dismiss();
        Timber.i("ON PAUSE IS CALLED FROM BALANCE");
    }

    private void setUpBalances() {
        initBalanceCard();
        balancesViewModel.getSelectedChannels().observe(getViewLifecycleOwner(), this::updateServices);
    }

    private void setUpLinkNewAccount() {
        addChannelLink = binding.newAccountLink;
        addChannelLink.setOnClickListener(v -> HomeFragment.navigateTo(Constants.NAV_LINK_ACCOUNT, requireActivity()));
    }

    private void initBalanceCard() {
        balanceTitle = binding.homeCardBalances.balanceHeaderTitleId;
        balanceTitle.setCompoundDrawablesRelativeWithIntrinsicBounds(balancesVisible ? R.drawable.ic_visibility_on : R.drawable.ic_visibility_off, 0, 0, 0);
        balanceTitle.setOnClickListener(view -> {
            showBalanceCards(!balancesVisible);
            showBubbleIfRequired();
        });

        balancesRecyclerView = binding.homeCardBalances.balancesRecyclerView;
        balancesRecyclerView.setLayoutManager(UIHelper.setMainLinearManagers(getContext()));
        balancesRecyclerView.setHasFixedSize(true);
    }

    private void showBalanceCards(boolean status) {
        toggleLink(status);
        balanceTitle.setCompoundDrawablesRelativeWithIntrinsicBounds(status ? R.drawable.ic_visibility_on : R.drawable.ic_visibility_off, 0, 0, 0);

        if (status) {
            balanceStack.setVisibility(GONE);
            binding.homeCardBalances.balancesMl.transitionToEnd();
        } else {
            balanceStack.setVisibility(VISIBLE);
            binding.homeCardBalances.balancesMl.transitionToStart();
        }

        balancesVisible = status;
    }

    private void updateServices(List<Channel> channels) {
        SHOW_ADD_ANOTHER_ACCOUNT = channels != null && !Channel.hasDummy(channels) && channels.size() > 1;
        addDummyChannelsIfRequired(channels);
        BalanceAdapter balanceAdapter = new BalanceAdapter(channels, (MainActivity) getActivity());
        balancesRecyclerView.setAdapter(balanceAdapter);
        balanceAdapter.showBalanceAmounts(true);

        showBalanceCards(Channel.areAllDummies(channels));
        updateStackCard(channels);

        channelList = channels;
        showBubbleIfRequired();
    }

    private void showBubbleIfRequired() {
        if(channelList !=null) {
            if(Channel.areAllDummies(channelList)) {
                if(!SHOWN_BUBBLE_MAIN_ACCOUNT && balancesVisible) {
                    Timber.i("SHOWING ALL DUMMIES");
                   firstAccountBubble = new ShowcaseExecutor(requireActivity(), binding).showcaseAddFirstAccount();
                    SHOWN_BUBBLE_MAIN_ACCOUNT = true;
                }
            }
            else if(Channel.hasDummy(channelList)) {
                if(!SHOWN_BUBBLE_OTHER_ACCOUNT && balancesVisible) {
                    Timber.i("SHOWING ONLY ONE DUMMIES");
                    SHOWN_BUBBLE_OTHER_ACCOUNT = true;
                    new Handler().postDelayed(() -> secondAccountBubble = new ShowcaseExecutor(requireActivity(), binding).showcaseAddSecondAccount(), 2000);
                }
            }
        }
    }

    private void updateStackCard(List<Channel> channels) {
        if(channels !=null) {
            List<Channel> tempChannels = new ArrayList<>(channels);
            Collections.reverse(tempChannels);

            BalanceCardStackAdapter cardStackAdapter = new BalanceCardStackAdapter(requireContext());
            balanceStack.setAdapter (cardStackAdapter);
            cardStackAdapter.updateData(tempChannels);
            balanceStack.setOverlapGaps(STACK_OVERLAY_GAP);
            balanceStack.setRotationX(ROTATE_UPSIDE_DOWN);
            balanceStack.setOnClickListener(view -> showBalanceCards(!balancesVisible));
            updateBalanceCardStackHeight(tempChannels.size());
        }
    }

    private void updateBalanceCardStackHeight(int numOfItems) {
        ViewGroup.LayoutParams params = balanceStack.getLayoutParams();
        params.height = 20 * numOfItems;
        balanceStack.setLayoutParams(params);
    }

    private void addDummyChannelsIfRequired(@Nullable List<Channel> channels) {
        if (channels != null && channels.size() == 0) {
            channels.add(new Channel().dummy(getString(R.string.your_main_account), GREEN_BG));
            channels.add(new Channel().dummy(getString(R.string.your_other_account), BLUE_BG));
        } else if (channels != null && channels.size() == 1)
            channels.add(new Channel().dummy(getString(R.string.your_other_account), BLUE_BG));
    }

    public void toggleLink(boolean show) {
        if(SHOW_ADD_ANOTHER_ACCOUNT) {
            addChannelLink.setVisibility(show ? VISIBLE : GONE);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }


}

File: app/src/main/java/com/hover/stax/balances/BalancesFragment.kt
package com.hover.stax.balances

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.cardview.widget.CardView
import androidx.core.content.ContentProviderCompat.requireContext
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.RecyclerView
import com.hover.sdk.actions.HoverAction
import com.hover.stax.MainNavigationDirections
import com.hover.stax.R
import com.hover.stax.accounts.Account
import com.hover.stax.accounts.AccountsViewModel
import com.hover.stax.accounts.DUMMY
import com.hover.stax.addChannels.ChannelsViewModel
import com.hover.stax.databinding.FragmentBalanceBinding
import com.hover.stax.home.HomeFragmentDirections
import com.hover.stax.home.MainActivity
import com.hover.stax.hover.AbstractHoverCallerActivity
import com.hover.stax.utils.AnalyticsUtil
import com.hover.stax.utils.UIHelper
import com.hover.stax.utils.collectLatestLifecycleFlow
import com.hover.stax.views.StaxDialog
import com.hover.stax.views.staxcardstack.StaxCardStackView
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import org.koin.androidx.viewmodel.ext.android.sharedViewModel
import org.koin.androidx.viewmodel.ext.android.viewModel


class BalancesFragment : Fragment(), BalanceAdapter.BalanceListener {

    private lateinit var addAccountBtn: CardView
    private lateinit var balanceTitle: TextView
    private lateinit var balanceStack: StaxCardStackView
    private lateinit var balancesRecyclerView: RecyclerView

    private var _binding: FragmentBalanceBinding? = null
    private val binding get() = _binding!!

    private val accountsViewModel: AccountsViewModel by sharedViewModel()
    private val balancesViewModel: BalancesViewModel by sharedViewModel()
    private val channelsViewModel: ChannelsViewModel by sharedViewModel()
    private lateinit var cardStackAdapter: BalanceCardStackAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentBalanceBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setUpBalances()
        setUpLinkNewAccount()
    }

    private fun setUpBalances() {
        setUpBalanceHeader()
        setUpBalanceList()
        setUpHiddenStack()

        balancesViewModel.showBalances.observe(viewLifecycleOwner) { showBalanceCards(it) }

        collectLatestLifecycleFlow(accountsViewModel.accounts) {
            updateAccounts(ArrayList(it))
        }

        collectLatestLifecycleFlow(balancesViewModel.balanceAction) {
            attemptCallHover(balancesViewModel.userRequestedBalanceAccount.value, it)
        }

        lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                channelsViewModel.accountCallback.collect {
                    Toast.makeText(requireActivity(), "Account ${it.name} event has been received", Toast.LENGTH_SHORT).show()
                    askToCheckBalance(it)
                }
            }
        }
    }

    private fun attemptCallHover(account: Account?, action: HoverAction?) {
        action?.let { account?.let { callHover(account, action) } }
    }

    private fun callHover(account: Account, action: HoverAction) {
        (requireActivity() as AbstractHoverCallerActivity).runSession(account, action)
    }

    private fun setUpLinkNewAccount() {
        addAccountBtn = binding.newAccountLink
        addAccountBtn.setOnClickListener {
            (requireActivity() as MainActivity).checkPermissionsAndNavigate(MainNavigationDirections.actionGlobalAddChannelsFragment())
        }
    }

    private fun setUpBalanceHeader() {
        balanceTitle = binding.homeCardBalances.balanceHeaderTitleId.also {
            it.setCompoundDrawablesRelativeWithIntrinsicBounds(
                if (balancesViewModel.showBalances.value == true) R.drawable.ic_visibility_on else R.drawable.ic_visibility_off,
                0,
                0,
                0
            )
            it.setOnClickListener { balancesViewModel.setBalanceState(!balancesViewModel.showBalances.value!!) }
        }
    }

    private fun setUpBalanceList() {
        balancesRecyclerView = binding.homeCardBalances.balancesRecyclerView.apply {
            layoutManager = UIHelper.setMainLinearManagers(context)
            setHasFixedSize(true)
            visibility = View.GONE
        }
    }

    private fun setUpHiddenStack() {
        cardStackAdapter = BalanceCardStackAdapter(requireActivity())
        balanceStack = binding.stackBalanceCards
        balanceStack.apply {
            setAdapter(cardStackAdapter)
            setOverlapGaps(STACK_OVERLAY_GAP)
            rotationX = ROTATE_UPSIDE_DOWN
        }
    }

    private fun showBalanceCards(show: Boolean) {
        AnalyticsUtil.logAnalyticsEvent(getString(if (balancesViewModel.showBalances.value != true) R.string.show_balances else R.string.hide_balances), requireActivity())
        balanceTitle.setCompoundDrawablesRelativeWithIntrinsicBounds(
            if (show) R.drawable.ic_visibility_on else R.drawable.ic_visibility_off, 0, 0, 0
        )

        showAddAccount(accountsViewModel.accounts.value, show)
        if (show) binding.homeCardBalances.balancesMl.transitionToEnd() else binding.homeCardBalances.balancesMl.transitionToStart()

        balanceStack.visibility = if (show) View.GONE else View.VISIBLE
    }

    private fun updateAccounts(accounts: ArrayList<Account>) {
        accounts.let {
            addDummyAccountsIfRequired(accounts)
            cardStackAdapter.updateData(accounts.reversed())
            updateBalanceCardStackHeight(accounts.size)
            showAddAccount(accounts, balancesViewModel.showBalances.value!!)
        }
        val balancesAdapter = BalanceAdapter(accounts, this)
        balancesRecyclerView.adapter = balancesAdapter
        showBalanceCards(balancesViewModel.showBalances.value!!)
    }

    private fun askToCheckBalance(account: Account) {
        val dialog = StaxDialog(requireActivity())
            .setDialogTitle(R.string.check_balance_title)
            .setDialogMessage(R.string.check_balance_desc)
            .setNegButton(R.string.later, null)
            .setPosButton(R.string.check_balance_title) { onTapRefresh(account) }
        dialog.showIt()
    }

    private fun updateBalanceCardStackHeight(numOfItems: Int) {
        val params = balanceStack.layoutParams
        params.height = 20 * numOfItems
        balanceStack.layoutParams = params
    }

    private fun showAddAccount(accounts: List<Account>?, show: Boolean) {
        addAccountBtn.visibility = if (!accounts.isNullOrEmpty() && accounts.size > 1 && show) View.VISIBLE else View.GONE
    }

    private fun addDummyAccountsIfRequired(accounts: ArrayList<Account>?) {
        accounts?.let {
            if (it.isEmpty()) {
                accounts.add(Account(getString(R.string.your_main_account), GREEN_BG).dummy())
                accounts.add(Account(getString(R.string.your_other_account), BLUE_BG).dummy())
            }
            if (it.size == 1)
                accounts.add(Account(getString(R.string.your_other_account), BLUE_BG).dummy())
        }
    }

    override fun onTapRefresh(account: Account?) {
        if (account == null || account.id == DUMMY)
            (requireActivity() as MainActivity).checkPermissionsAndNavigate(HomeFragmentDirections.actionNavigationHomeToNavigationLinkAccount())
        else {
            AnalyticsUtil.logAnalyticsEvent(getString(R.string.refresh_balance_single), requireContext())
            balancesViewModel.requestBalance(account)
        }
    }

    override fun onTapDetail(accountId: Int) {
        if (accountId == DUMMY)
            (requireActivity() as MainActivity).checkPermissionsAndNavigate(HomeFragmentDirections.actionNavigationHomeToNavigationLinkAccount())
        else
            findNavController().navigate(HomeFragmentDirections.actionNavigationHomeToAccountDetailsFragment(accountId))
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        const val GREEN_BG = "#46E6CC"
        const val BLUE_BG = "#04CCFC"

        const val STACK_OVERLAY_GAP = 10
        const val ROTATE_UPSIDE_DOWN = 180f
        const val BALANCE_VISIBILITY_KEY: String = "BALANCE_VISIBLE"
    }
}